### PR TITLE
[ConstraintSystem] Respect the constraint solver performance thresholds,

### PR DESCRIPTION
**Pull request by hborla**
including time and allocated memory, in mergePartialSolutions.

Previously, the solver could get stuck in `mergePartialSolutions` in the do/while loop and never recover because it didn't consider the performance thresholds.

Resolves: rdar://problem/53589951
Resolves: rdar://problem/60062211

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -4509,6 +4506,14 @@ class ConstraintSystem {
     return false;
   }
 
+  bool getExpressionTooComplex(SmallVectorImpl<Solution> const &solutions) {
+    size_t solutionMemory = 0;
+    for (auto const& s : solutions) {
```
Comment (xedin) [inline lib/Sema/ConstraintSystem.h]: Nit: since like a good use-case for `std::accumulate`.

**Review by xedin — APPROVED** [review]
LGTM!

File: lib/Sema/ConstraintSystem.h (hunk 1 above)
Comment (hborla) [reply lib/Sema/ConstraintSystem.h]: I think using `std::accumulate` makes it harder to read:
```cpp
    size_t solutionMemory =
        std::accumulate(solutions.begin(), solutions.end(), 0,
            [](size_t memory, const Solution &solution) -> size_t {
              return memory + solution.getTotalMemory();
            });
    return getExpressionTooComplex(solutionMemory);
```
or
```cpp
    return getExpressionTooComplex(
        std::accumulate(solutions.begin(), solutions.end(), 0,
            [](size_t memory, const Solution &s) -> size_t {
              return memory + s.getTotalMemory();
            }));
```
What do you think?

Comment (xedin) [reply lib/Sema/ConstraintSystem.h]: I'd prefer #2 but it really doesn't matter that much :)

**Pull request merged**